units/sleep_scheduler.go

```diff
@@ -0,0 +1,129 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const sleepSchedulerJobName = "sleep-scheduler"
+
+func init() {
+	registry.AddJobType(sleepSchedulerJobName, func() amboy.Job {
+		return makeSleepSchedulerJob()
+	})
+}
+
+type sleepSchedulerJob struct {
+	job.Base `bson:"job_base" json:"job_base" yaml:"job_base"`
+	env      evergreen.Environment
+}
+
+// NewSleepSchedulerJob creates a job to manage unexpirable host sleep
+// schedules.
+func NewSleepSchedulerJob(env evergreen.Environment, ts string) amboy.Job {
+	j := makeSleepSchedulerJob()
+	j.SetID(fmt.Sprintf("%s.%s", sleepSchedulerJobName, ts))
+	j.env = env
+	j.SetScopes([]string{sleepSchedulerJobName})
+	return j
+}
+
+func makeSleepSchedulerJob() *sleepSchedulerJob {
+	j := &sleepSchedulerJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    sleepSchedulerJobName,
+				Version: 1,
+			},
+		},
+	}
+	return j
+}
+
+func (j *sleepSchedulerJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+	if err := j.populateIfUnset(); err != nil {
+		j.AddError(err)
+		return
+	}
+
+	ts := utility.RoundPartOfMinute(0)
+	if err := populateQueueGroup(ctx, j.env, spawnHostModificationQueueGroup, j.makeStopAndStartJobs, ts); err != nil {
+		j.AddError(errors.Wrap(err, "enqueuing stop and start jobs"))
+		return
+	}
+}
+
+func (j *sleepSchedulerJob) populateIfUnset() error {
+	if j.env == nil {
+		j.env = evergreen.GetEnvironment()
+	}
+	return nil
+}
+
+func (j *sleepSchedulerJob) makeStopAndStartJobs(ctx context.Context, _ evergreen.Environment, ts time.Time) ([]amboy.Job, error) {
```

Comment:
I'm surprised we're doing this as part of its own job as opposed to looping directly in sleepScheduleJobs or having a populate function. What's the thought behind having a job to do the jobs, does the query seem expensive?

Reply:
Doing it in the cron populator is probably fine in terms of query performance because it's not that expensive and the minutely cron populator only runs once per minute (on one out of all the app servers). Since the cron has to run once per minute though, it would be nice to avoid overloading the minutely cron with too many queries if possible, since each query we add to the cron populator will slow it down.

The other benefit is that the code organization will be a little easier for future changes to sleep schedule logic. The job also has do some more complex stuff in later tickets to manage exemptions, so splitting out the logic now would be useful.